### PR TITLE
Fix SymDB upload dropped requests

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do
SymbolSink was emitting an upload requests for every scope added into the sink queue. As the queue was 1024 entries in capacity we could easily end up in a situation where more than 20 requests are made when draining the Q. to avoid that we are batching the scope together in a single larger upload request and ensure that the sink queue is flushed when full before adding new element and not losing any symbols.

# Motivation
avoid dropping SymDB requests

# Additional Notes

# Contributor Checklist

- [ ] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [ ] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [ ] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [ ] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [ ] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [DEBUG-2685]

[DEBUG-2685]: https://datadoghq.atlassian.net/browse/DEBUG-2685?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java [hunk 1]
```diff
@@ -59,37 +58,51 @@ public void stop() {
...
   }
 
   public void flush() {
     if (scopes.isEmpty()) {
       return;
     }
     List<ServiceVersion> scopesToSerialize = new ArrayList<>();
+    // ArrayBlockingQueue makes drainTo atomic, so it is safe to call flush from different and
+    // concurrent threads
     scopes.drainTo(scopesToSerialize);
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java:84]: maybe we should check scopesToSerialize is empty just return

**Review by shatzi — APPROVED** [review]

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java [hunk 2]
```diff
@@ -59,37 +58,51 @@ public void stop() {
...
-        stats.updateStats(classScopeCount, json.length());
-        symbolUploader.uploadAsMultipart(
-            "",
-            event,
-            new BatchUploader.MultiPartContent(
-                json.getBytes(StandardCharsets.UTF_8), "file", "file.json"));
-      } catch (Exception e) {
-        ExceptionHelper.logException(LOGGER, e, "Error during scope serialization:");
-      }
+      allScopes.addAll(serviceVersion.getScopes());
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java:87]: removing ServiceVersion make it simpler.
simplify by adding the getScopeCount method, but updateStats method stays

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java [hunk 3]
```diff
@@ -59,37 +58,51 @@ public void stop() {
...
-  public boolean addScope(Scope jarScope) {
+  public void addScope(Scope jarScope) {
     ServiceVersion serviceVersion =
         new ServiceVersion(serviceName, env, version, "JAVA", Collections.singletonList(jarScope));
-    return scopes.offer(serviceVersion);
+    boolean added = scopes.offer(serviceVersion);
+    int retries = 10;
+    while (!added) {
+      // Q is full, flushing synchronously
+      flush();
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java:68]: the flush make upload http requests, but http requests are sent async in OkHttp component.

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java (hunk 1 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java:84]: done

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java [hunk 4]
```diff
@@ -59,37 +57,47 @@ public void stop() {
     symbolUploader.shutdown();
   }
 
-  public boolean addScope(Scope jarScope) {
-    ServiceVersion serviceVersion =
-        new ServiceVersion(serviceName, env, version, "JAVA", Collections.singletonList(jarScope));
-    return scopes.offer(serviceVersion);
+  public void addScope(Scope jarScope) {
+    boolean added = scopes.offer(jarScope);
+    int retries = 10;
```
Comment (ojung) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java:62]: Why might we need to retry more than once if `flush()` is synchronous?

**Review by ojung — COMMENTED** [review]
The way I understand these changes, we could still end up with more than 20 concurrent requests: When there are (20 + 1) * 1024 classes loaded at roughly the same time. Please correct me if I'm wrong or this isn't a concern.

**Comment by jpbempel** [comment]
> The way I understand these changes, we could still end up with more than 20 concurrent requests: When there are (20 + 1) * 1024 classes loaded at roughly the same time. Please correct me if I'm wrong or this isn't a concern.

That's almost true, except that top level scopes are aggregated by jar name, en so each entry in the Queue is a different jar. So you need (20+1) *1024 different jars for that :)

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java (hunk 4 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/sink/SymbolSink.java:62]: concurrency at the best: this flushing process is in background thread, and still (but unlikely) threads are loading a lot of classes/jars and we filling up again the Q while flushing the previous batch, we are trying again to flush.

Flush is synchronous, but we are still allowing to add to the queue once drained.

**Pull request merged**